### PR TITLE
Feature/gh 13806 exit button

### LINKED ISSUE #13806: Create new component for "EXIT CONFIGURATION" Button in SPA Header
New component should trigger a "browser back" when clicked. 
When no browser back is possible, the user should go the the product detail of the product which is configured.

Component will be added in rulebased module, lazy loading is expected.
Component has to take care that the header is styled correctly.

### PR DESCRIPTION
**Pull request by Changsuwan**

### CODE REVIEW

File: feature-libs/product-configurator/common/assets/translations/en/configurator-common.ts [hunk 1]
```diff
@@ -54,6 +54,8 @@ export const configurator = {
       select: 'Select',
       add: 'Add',
       remove: 'Remove',
+      exit: 'Exit Configuration',
+      exitshort: 'Exit',
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/common/assets/translations/en/configurator-common.ts]: Please, use a camel case, for instance `exitShort` or better `exitMobil`

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<button class="cx-config-exit-button" tabindex="0" (click)="goBack()">
+  <span class="cx-config-exit-button-text">
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html]: It is necessary  to wrap text within the button with a `span`
Please, remove it

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  constructor(
+    protected productService: ProductService,
+    protected routingService: RoutingService,
+    protected configRouterExtractorService: ConfiguratorRouterExtractorService,
+    protected configuratorCommonsService: ConfiguratorCommonsService,
+    protected breakpointService: BreakpointService,
+    protected windowRef: WindowRef
+  ) {}
+
+  goBack() {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts]: Please, write a comment

File: projects/storefrontstyles/scss/components/layout/header/_header.scss [hunk 4]
```diff
@@ -87,6 +87,34 @@ $space: 0.5rem;
     }
   }
 
+  .VariantConfigExitButton {
+    margin-top: 14px;
```
Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: `margin-block-start` should be used instead `margin-top`

File: projects/storefrontstyles/scss/components/layout/header/_header.scss [hunk 5]
```diff
@@ -87,6 +87,34 @@ $space: 0.5rem;
...
+    margin-bottom: 14px;
+    margin-inline-start: auto;
+    margin-inline-end: 14px;
+
+    @include media-breakpoint-down(sm) {
+      margin-top: 4px;
```
Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: `margin-block-start` should be used instead `margin-top`

File: projects/storefrontstyles/scss/components/layout/header/_header.scss [hunk 6]
```diff
@@ -87,6 +87,34 @@ $space: 0.5rem;
...
+      margin-inline-end: 0;
+      margin-bottom: 4px;
+      margin-inline-start: auto;
+    }
+  }
+
+  .VariantConfigOverviewExitButton {
+    margin-top: 14px;
```
Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: `margin-block-start` should be used instead `margin-top`

File: projects/storefrontstyles/scss/components/layout/header/_header.scss [hunk 7]
```diff
@@ -87,6 +87,34 @@ $space: 0.5rem;
     }
   }
 
+  .VariantConfigExitButton {
+    margin-top: 14px;
+    margin-bottom: 14px;
```
Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: `margin-block-end` should be used instead `margin-bottom`

File: projects/storefrontstyles/scss/components/layout/header/_header.scss [hunk 8]
```diff
@@ -87,6 +87,34 @@ $space: 0.5rem;
...
+    margin-inline-start: auto;
+    margin-inline-end: 14px;
+
+    @include media-breakpoint-down(sm) {
+      margin-top: 4px;
+      margin-inline-end: 0;
+      margin-bottom: 4px;
```
Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: `margin-block-end` should be used instead `margin-bottom`

File: projects/storefrontstyles/scss/components/layout/header/_header.scss [hunk 9]
```diff
@@ -87,6 +87,34 @@ $space: 0.5rem;
...
+      margin-inline-start: auto;
+    }
+  }
+
+  .VariantConfigOverviewExitButton {
+    margin-top: 14px;
+    margin-bottom: 14px;
```
Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: `margin-block-end` should be used instead `margin-bottom`

File: projects/storefrontstyles/scss/components/layout/header/_header.scss [hunk 10]
```diff
@@ -87,6 +87,34 @@ $space: 0.5rem;
...
+    margin-inline-start: auto;
+    margin-inline-end: 14px;
+
+    @include media-breakpoint-down(sm) {
+      margin-top: 4px;
+      margin-inline-end: 0;
+      margin-bottom: 4px;
```
Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: `margin-block-end` should be used instead `margin-bottom`

File: projects/storefrontstyles/scss/components/layout/header/_header.scss [hunk 11]
```diff
@@ -87,6 +87,34 @@ $space: 0.5rem;
...
+  }
+
+  .VariantConfigOverviewExitButton {
+    margin-top: 14px;
+    margin-bottom: 14px;
+    margin-inline-start: auto;
+    margin-inline-end: 14px;
+
+    @include media-breakpoint-down(sm) {
+      margin-top: 4px;
```
Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: `margin-block-start` should be used instead `margin-top`

File: projects/storefrontstyles/scss/components/layout/header/_header.scss [hunk 12]
```diff
@@ -87,6 +87,34 @@ $space: 0.5rem;
     }
   }
 
+  .VariantConfigExitButton {
```
Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: Try to avoid code duplication. You could work with the comma operator

Comment (Larisa-Staroverova) [inline projects/storefrontstyles/scss/components/layout/header/_header.scss]: The better solution without code duplication could be:
```
 .VariantConfigExitButton,
  .VariantConfigOverviewExitButton {
    margin-block-start: 14px;
    margin-block-end: 14px;
    margin-inline-start: auto;
    margin-inline-end: 14px;

    @include media-breakpoint-down(sm) {
      margin-block-start: 4px;
      margin-inline-end: 0px;
      margin-block-end: 4px;
      margin-inline-start: auto;
    }
  }
```

File: feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss [hunk 13]
```diff
@@ -0,0 +1,23 @@
+%cx-configurator-exit-button {
+  .cx-config-exit-button {
+    border-radius: var(--cx-buttons-border-radius);
+    border: 2px solid white;
+    background-color: transparent;
+    height: 44px;
+    color: white;
+  }
+  .cx-exit-configurator {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss]: Code duplication!

File: feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss [hunk 14]
```diff
@@ -0,0 +1,23 @@
+%cx-configurator-exit-button {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss:1]: See this working solution:
```
%cx-configurator-exit-button {
  align-items: center;
  font-weight: 600;
  font-size: 15px;
  margin-inline-start: 5px;
  margin-inline-end: 5px;

  button {
    border-radius: var(--cx-buttons-border-radius);
    border: 2px solid white;
    background-color: transparent;
    height: 44px;
    color: white;
  }
}
```

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html [hunk 15]
```diff
@@ -0,0 +1,10 @@
+<button class="cx-config-exit-button" tabindex="0" (click)="goBack()">
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html]: Shorter solution:
```
<button tabindex="0" (click)="goBack()">
    <ng-container *ngIf="isDesktop() | async">
      {{ 'configurator.button.exit' | cxTranslate }}
    </ng-container>
    <ng-container *ngIf="isMobile() | async">
      {{ 'configurator.button.exitshort' | cxTranslate }}
    </ng-container>
</button>
```

**Review by Larisa-Staroverova — COMMENTED** [review]
Please, see my comments

File: projects/storefrontstyles/scss/components/layout/header/_header.scss (hunk 6 above)
Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: done

File: projects/storefrontstyles/scss/components/layout/header/_header.scss (hunk 9 above)
Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: done

File: projects/storefrontstyles/scss/components/layout/header/_header.scss (hunk 11 above)
Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: done

File: projects/storefrontstyles/scss/components/layout/header/_header.scss (hunk 10 above)
Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: done

File: projects/storefrontstyles/scss/components/layout/header/_header.scss (hunk 8 above)
Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: done

File: projects/storefrontstyles/scss/components/layout/header/_header.scss (hunk 5 above)
Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: done

File: projects/storefrontstyles/scss/components/layout/header/_header.scss (hunk 7 above)
Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: done

File: projects/storefrontstyles/scss/components/layout/header/_header.scss (hunk 4 above)
Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: done

File: projects/storefrontstyles/scss/components/layout/header/_header.scss (hunk 12 above)
Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: Done

Comment (Changsuwan) [reply projects/storefrontstyles/scss/components/layout/header/_header.scss]: Done

File: feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss (hunk 13 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss]: Fixed

File: feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss (hunk 14 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss:1]: Done

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts (hunk 3 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts]: Done

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html (hunk 2 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html]: Done

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html (hunk 15 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html]: Done

File: feature-libs/product-configurator/common/assets/translations/en/configurator-common.ts (hunk 1 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/common/assets/translations/en/configurator-common.ts]: Done

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html (hunk 15 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.html]: Please, remember to **Resolve conversation**, when you are done with the comment. Thanks!

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts (hunk 3 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts]: Please, remember to Resolve conversation, when you are done with the comment. Thanks!

File: feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss (hunk 14 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/styles/_configurator-exit-button.scss:1]: Please, remember to **Resolve conversation**, when you are done with the comment. Thanks!

**Review by Larisa-Staroverova — CHANGES_REQUESTED** [review]
`goBack` method has not been tested yet. 2 tests are failing. Please, fix them!

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    protected configuratorCommonsService: ConfiguratorCommonsService,
+    protected breakpointService: BreakpointService,
+    protected windowRef: WindowRef
+  ) {}
+  /**
+   * the method navigates to the previous page when user clicks on the configurator exit button.
+   * If there is no previous page to return to, it will navigate to the product detail page of the product
+   * that is being configured.
+   */
+  goBack() {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts]: Please, rename this method, for instance `exit` or `exitConfiguration`
The current method name does not correspond to what this method does, namely exit a configuration

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+  constructor(
+    protected productService: ProductService,
+    protected routingService: RoutingService,
+    protected configRouterExtractorService: ConfiguratorRouterExtractorService,
+    protected configuratorCommonsService: ConfiguratorCommonsService,
+    protected breakpointService: BreakpointService,
+    protected windowRef: WindowRef
+  ) {}
+  /**
+   * the method navigates to the previous page when user clicks on the configurator exit button.
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts]: The comment should start with a verb in 3-d person singular, for instance `navigates`, `exits`
Furthermore, your comment does not correspond to what this method does, namely exit a configuration.
Please, adjust the comment accordingly.

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.spec.ts [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+    );
+    htmlElem = fixture.nativeElement;
+    fixture.detectChanges();
+  });
+
+  it('should create component', () => {
+    expect(component).toBeDefined();
+  });
+
+  describe('navigation tests', () => {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.spec.ts]: Please, rename the test description, for instance `exit a configuration`

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+    it('should navigate to product detail page if going back to previous page does not work', () => {
```
Comment (Larisa-Staroverova) [inline feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.spec.ts]: According to the current implementation, after clicking on the `exit configuration` button, the end-user always exits a configuration / goes to the product detail page.
Please, adjust the test description accordingly.

**Review by Larisa-Staroverova — CHANGES_REQUESTED** [review]
Please, take into account my comments

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts (hunk 16 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts]: Done

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts (hunk 17 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.ts]: Done

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.spec.ts (hunk 19 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.spec.ts]: Done

File: feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.spec.ts (hunk 18 above)
Comment (Changsuwan) [reply feature-libs/product-configurator/rulebased/components/exit-button/configurator-exit-button.component.spec.ts]: Done

**Review by Larisa-Staroverova — DISMISSED** [review]

**Review by Larisa-Staroverova — APPROVED** [review]

**Pull request merged**